### PR TITLE
Slice/Span performance improvements

### PR DESCRIPTION
**Pull request by adamsitnik**
The optimizations I have introduced:
1. Making Length a readonly field instead of property. It gives a lot for the "for loop"
2. Simpler range check by casting index and length to unsigned integers to eliminate index >= 0 check
3. Removing Dispose from struct Enumerator to avoid compiler from putting foreach in try+finally block
4. Introducing new EnumeratorObject class that is created when the collection is being enumerated via the IEnumerable interface. Since IEnumerable.GetEnumerator() returns IEnumerator (interface) it would be costly if this was struct because invoking interface method on structures is expensive
5. Added inlining of indexer getter, which makes the EnumeratorObject inline the indexer method.
6. Added GetItemWithoutBoundariesCheck method, which is used in safe EnumeratorObject, where Improved MoveNext ensures that it is impossible to move outside boundaries. It eliminates single range check (stadard indexer)
7. Equals based on for loop which reduces number of method calls (enumerating via Enumerator calls to MoveNext and Current, looping just to indexer) and added generic Comparer to avoid boxing (but I had to add dependency to System.Collections to use EqualityComparer<T>.Default)

I have tested all these changes a long time ago when I added them to the original Slice repo of Joe Duffy

### CODE REVIEW

**Comment by KrzysztofCwalina** [comment]
Dzieki!
.

**Comment by KrzysztofCwalina** [comment]
oops. closed by mistake.

**Comment by VSadov** [comment]
I have been reviewing these changes in the original Slice repo and was waiting for reappearance here.
Good stuff !!
LGTM

File: src/System.Slices/src/System/SpanEnumerators.cs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            return new EnumeratorObject(this);
+        }
+
+        /// <summary>
+        /// A struct-based enumerator, to make fast enumerations possible.
+        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// </summary>
+        public struct Enumerator
+        {
```
Comment (VSadov) [inline src/System.Slices/src/System/SpanEnumerators.cs]: The nice part here is that foreach -pattern iterator does not implement IDisposable, so C# codegen for foreach will be much better.
Otherwise compiler would need to wrap the whole deal in try/finally to guarantee that Dispose is called. And that in turn would impact quality of JITed code as well.

File: src/System.Slices/src/System/Contract.cs [hunk 2]
```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
...
         public static void RequiresInRange(int start, int length)
         {
-            if (!(start >= 0 && start < length))
+            // explicit unchecked block is used in order to avoid OverflowException when casting 
+            // negative int to unsigned int when the code is compiled with /checked option
+            unchecked
             {
-                throw new ArgumentOutOfRangeException();
+                // use unsigned int to reduce range check by one ( >= 0 )
+                if ((uint)start >= (uint)length)
```
Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: Does this do what you think it does? What is the value of `u` in the following? `int i = -1; uint u = (uint)i;` I'd assume `u == Uint32.MaxValue`.

Comment (VSadov) [inline src/System.Slices/src/System/Contract.cs]: Seems like the helper assumes nonnegative length. It makes sense to use it in checks like index range checking  (since we know the span length is nonnegative by construction). In fact this is exactly the same technique as used by JIT for array index operations.
Perhaps it is worth adding Debug.Assert(length >= 0) ?

Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: Asserting the value is the least we should be doing here, but asserts are generally compiled out and this is a `public` entry-point. Seems dangerous to me. I'm not sure you can get away from a two part test scenario. `length >= 0 && (uint)start < (uint)length`.

There are many days when I wish CLS didn't care about signed-ness.

Comment (VSadov) [inline src/System.Slices/src/System/Contract.cs]: I think Contract class is not public API.

For a public entry point Assert would not be sufficient indeed. In such case it would be worth having a separate private helper specifically for cases where we know for a fact that length is nonnegative. It is a very common case.

Comment (adamsitnik) [inline src/System.Slices/src/System/Contract.cs]: I did some calculations. The minimal value that can be created out of casting signed negative integer to unsigned integer is 2,147,483,648. .NET has limit for single dimension array size which is 2,147,483,591 for bytes array and 2,146,435,071  for array of other types.

So the minimal value that can be generated is always bigger that the .NET's array size limit so the single comparision with Length should be always safe. 

What do you guys think?

Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: @adamsitnik is that a guaranteed limit forever on all platforms and version of the framework? If not, you're trading a large amount of risk for a very minimal performance gain.

Comment (adamsitnik) [inline src/System.Slices/src/System/Contract.cs]: @whoisj The gain is small, but the indexer is probably going to be the "hottest path" here.

IMHO the size limit is so old that probably a lot of system parts rely on it and it is not going to change in the future. If it changes, the whole idea of Slice is to have small pieces of huge vector, so the Length should be much smaller than Max, because it makes no sense otherwise. And even if somebody abuses that in the future he or she is going to get some item from the collection instead of Exception so we keep the Memory Safety. 

List<T> also relies on that: http://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,0c65bec4d3fac21e

I think that we can keep it, but if you wish I can revert this change easily.

Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: I'm not the final call here - I'm just helping highlight potential issues :smile: 

Thanks for doing the homework - and I agree, every cycle matters on the hot path.

Can we get unit tests to prove this is the correct decision?

Comment (adamsitnik) [inline src/System.Slices/src/System/Contract.cs]: @whoisj I am glad that I could help and learn a lot at the same time. Just added the tests. Do they convince you?

File: src/System.Slices/src/System/Span.cs [hunk 3]
```diff
@@ -333,63 +323,15 @@ public override bool Equals(object obj)
...
-        /// <summary>
-        /// A struct-based enumerator, to make fast enumerations possible.
-        /// This isn't designed for direct use, instead see GetEnumerator.
+        /// Returns item from given index without the boundaries check
+        /// use only in places where moving outside the boundaries is impossible
+        /// gain: performance: no boundaries check (single operation) 
         /// </summary>
-        public struct Enumerator : IEnumerator<T>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T GetItemWithoutBoundariesCheck(int index)
```
Comment (krwq) [inline src/System.Slices/src/System/Span.cs]: Should this be public + unsafe? We don't want people to do any hacks to achieve the same thing

File: src/System.Slices/src/System/SpanEnumerators.cs (hunk 1 above)
Comment (krwq) [inline src/System.Slices/src/System/SpanEnumerators.cs]: This is really nice! Does compiler do that even if you leave the dispose empty and not get the results through IEnumerator?

Comment (mburbea) [inline src/System.Slices/src/System/SpanEnumerators.cs]: I think you don't need two separate enumerators. Can you not just implement the `IDisposable.Dispose` as an explicit implementation? Wouldn't that cause roslyn not to generate the call when it uses pattern matching?

Comment (VSadov) [inline src/System.Slices/src/System/SpanEnumerators.cs]: Once compiler sees that enumerator implements IDisposable it will be required to emit try/finally and call Dispose in there. Compiler cannot make assumptions about Dispose having no sideeffects.

Comment (VSadov) [inline src/System.Slices/src/System/SpanEnumerators.cs]: The other subtle thing here is that IEnumerator has stronger requirements for MoveNext. Implementation must continue returning "false" from MoveNext after iteration is over, no matter how many times called. (think of the overflowing the index). 
Foreach pattern does not impose such requirements. Since compiler will not call MoveNext redundantly, it is possible to have simpler and more efficient implementation.

If performance is super important, and it is in this case since we are dealing with fundamental scenario competing with array indexing, it would be reasonable to have separate enumerators.

Comment (adamsitnik) [inline src/System.Slices/src/System/SpanEnumerators.cs]: I have written some benchmarks to show why having two dedicated enumerators is better for performance. The code can be found here: https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/Collections/EnumeratorsBenchmarks.cs

The results (scroll down to summary): https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/results/EnumeratorsBenchmarks.txt

Summary:
1. Calling interface method on struct enumerators is very expensive. Looping with interface method calls compared to direct calls is x3 times slower on x86 and more than 10x slower on x64.
2.  On x86 looping with direct method calls to class enumerator is 15% faster than to direct with struct. But on the other hand on x64 the looping with direct calls to struct enumerator is more than 3x times faster than class.

That is why we have two dedicated enumerators here.

File: src/System.Slices/src/System/Contract.cs (hunk 2 above)
Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: If the maintainers are fine, I'm fine. :wink:

Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Contract.cs]: @adamsitnik, do you have any performance comparisons of the two alternatives (with and without >0 check)?

File: src/System.Slices/src/System/Span.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: Let's wait with exposing this publicly. Unsafe on a method with no pointers in signature can be called in safe context, and the method when used incorrectly could violate memory safety.

File: src/System.Slices/src/System/SpanEnumerators.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/SpanEnumerators.cs]: I was looking at the tests and the results and I don't seem to see a difference. I am looking at SingleStructEnumeratorCalledDirectly and DedicatedStructEnumeratorCalledDirectly  and they seem to be roughly equal in performance (similar ns/op)

Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/SpanEnumerators.cs]: OK, I think I see it. You are comparing SingleStructEnumeratorCalledViaInterface vs DedicatedClassEnumeratorCalledViaInterface.

File: src/System.Slices/src/System/Contract.cs [hunk 4]
```diff
@@ -31,17 +31,26 @@ public static void Requires(bool condition)
...
+                {
+                    throw new ArgumentOutOfRangeException();
```
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Contract.cs]: Could you pass the arg to the ctor?

File: src/System.Slices/src/System/Contract.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Contract.cs]: BTW, from practical perspective, I would add the check for >0 and do a commit. Then create a separate PR to remove the check if we see the benefit/risk being large.

File: src/System.Slices/src/System/Span.cs [hunk 5]
```diff
@@ -13,14 +12,14 @@ namespace System
...
-    public struct Span<T> : IEnumerable<T>, IEquatable<Span<T>>
+    public partial struct Span<T> : IEnumerable<T>, IEquatable<Span<T>>
     {
         /// <summary>A managed array/string; or null for native ptrs.</summary>
         readonly object _object;
         /// <summary>An byte-offset into the array/string; or a native ptr.</summary>
         readonly UIntPtr _offset;
         /// <summary>Fetches the number of elements this Span contains.</summary>
-        public int Length { get; private set; }
+        public readonly int Length;
```
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: Can you get the same benefit by making the filed readonly and then wrapping in inlinable Length property?

File: src/System.Slices/src/System/Span.cs [hunk 6]
```diff
@@ -299,27 +299,17 @@ public bool Equals(Span<T> other)
                 return false;
             }
 
-            Enumerator thisIt = this.GetEnumerator();
-            Enumerator otherIt = other.GetEnumerator();
-            while (true)
+            var nonBoxingComparer = EqualityComparer<T>.Default;
```
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: @krwq (cc: @adamsitnik), this is not blocking this commit, but could we optimize this even more by simply comparing memory? Walking every single byte in a Span<byte> (probably the main use of Span<T>) seems expensive in comparison.

**Comment by KrzysztofCwalina** [comment]
Nice changes. As I said in the relevant comment, I would just do a full range check and then do a separate PR later. And so besides this and a couple of small nits, it looks very good. 
Thanks!

File: src/System.Slices/src/System/SpanEnumerators.cs (hunk 1 above)
Comment (VSadov) [inline src/System.Slices/src/System/SpanEnumerators.cs]: The issue is mostly that once C# sees IDisposable (which unavoidably comes with IEnumerator), it will have to wrap the whole thing into "using". 
It impacts performance of foreach enough that it makes sense to have a separate light-weight struct enumerator just for foreach pattern.

Then, once you have two iterators, the one that implements IEnumerator is better be a class because you have to allocate anyways (new or box) and interface dispatch on boxed structs is noticeable slower.

We came to exactly same conclusions when implementing ImmutableArray.Enumerator
https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray%601%2BEnumerator.cs

Comment (VSadov) [inline src/System.Slices/src/System/SpanEnumerators.cs]: In ideal world, JIT would allocate a class iterator on the stack when possible (it seems possible in typical foreach), then devirtualize call to Dispose and notice that Dispose is actually empty and thus erase it as well as redundant try/finally.

In such case we could just have one enumerator that is a class. This is not the case today. :-(

File: src/System.Slices/src/System/Contract.cs (hunk 2 above)
Comment (whoisj) [inline src/System.Slices/src/System/Contract.cs]: > BTW, from practical perspective, I would add the check for >0 and do a commit. Then create a separate PR to remove the check if we see the benefit/risk being large.

:+1:

File: src/System.Slices/src/System/Span.cs (hunk 6 above)
Comment (adamsitnik) [inline src/System.Slices/src/System/Span.cs]: @KrzysztofCwalina @krwq I think that it is unusual to have collection's Equal and GetHashCode method comparing and being dependent on the content. Maybe we should implement some SequenceEqual extension methods? One with the memory comparision for structs and other with IEqutable<T> constraint?

Comment (krwq) [inline src/System.Slices/src/System/Span.cs]: It feels like it should exist on the span. I feel like this type without any useful methods will be quite useless. I'm not convinced about implementing all the interfaces although memory comparison, enumerators etc. should be here IMO

File: src/System.Slices/src/System/Span.cs (hunk 5 above)
Comment (adamsitnik) [inline src/System.Slices/src/System/Span.cs]: @KrzysztofCwalina  Unfortunately not, link to [benchmark](https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/Collections/LengthForLoopBenchmarks.cs) and [results](https://github.com/adamsitnik/DemoCode/blob/master/Performance/Benchmarks/results/LengthForLoopBenchmarks.txt)

File: src/System.Slices/src/System/Contract.cs (hunk 4 above)
Comment (adamsitnik) [inline src/System.Slices/src/System/Contract.cs]: I initially wanted to do this, but the problem is that this is an internal heper method, that might be called from public methods with different parameter names (start & index).

File: src/System.Slices/src/System/Contract.cs (hunk 2 above)
Comment (adamsitnik) [inline src/System.Slices/src/System/Contract.cs]: reverted, will be back in separate PR ;)

**Comment by adamsitnik** [comment]
@KrzysztofCwalina I have squashed the commits, revert the problematic changes and fixed the compilation error that I had introduced before.

File: src/System.Slices/src/System/Span.cs (hunk 5 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/src/System/Span.cs]: cc @krzysztofcwalina, it would be great to investigate why. But ok, let's use as a field first.
